anvil/src/udev.rs

```diff
@@ -700,6 +709,119 @@ impl AnvilState<UdevData> {
+            o.user_data().get::<UdevOutputId>()
+                == Some(&UdevOutputId {
+                    device_id: surface.device_id,
+                    crtc,
+                })
+        }) {
+            output.clone()
+        } else {
+            // somehow we got called with an invalid output
+            return;
+        };
+
+        let schedule_render = match surface
+            .surface
+            .frame_submitted()
+            .map_err(Into::<SwapBuffersError>::into)
+        {
+            Ok(user_data) => {
+                if let Some(mut feedback) = user_data.flatten() {
+                    let tp = metadata.as_ref().and_then(|metadata| match metadata.time {
+                        smithay::backend::drm::DrmEventTime::Monotonic(tp) => Some(tp),
+                        smithay::backend::drm::DrmEventTime::Realtime(_) => None,
+                    });
+                    let seq = metadata.as_ref().map(|metadata| metadata.sequence).unwrap_or(0);
+
+                    let (clock, flags) = if let Some(tp) = tp {
+                        (
+                            tp.into(),
+                            wp_presentation_feedback::Kind::Vsync
+                                | wp_presentation_feedback::Kind::HwClock
+                                | wp_presentation_feedback::Kind::HwCompletion,
+                        )
+                    } else {
+                        (self.clock.now(), wp_presentation_feedback::Kind::Vsync)
+                    };
+
+                    feedback.presented(
+                        clock,
+                        output
+                            .current_mode()
+                            .map(|mode| mode.refresh as u32)
+                            .unwrap_or_default(),
+                        seq as u64,
+                        flags,
+                    );
+                }
+
+                true
+            }
+            Err(err) => {
+                warn!(self.log, "Error during rendering: {:?}", err);
+                match err {
+                    SwapBuffersError::AlreadySwapped => true,
+                    SwapBuffersError::TemporaryFailure(err) => matches!(
+                        err.downcast_ref::<DrmError>(),
+                        Some(&DrmError::DeviceInactive)
+                            | Some(&DrmError::Access {
+                                source: drm::SystemError::PermissionDenied,
+                                ..
+                            })
+                    ),
+                    SwapBuffersError::ContextLost(err) => panic!("Rendering loop lost: {}", err),
+                }
+            }
+        };
+
+        if schedule_render {
+            let output_refresh = match output.current_mode() {
+                Some(mode) => mode.refresh,
+                None => return,
+            };
+            let repaint_delay =
+                Duration::from_millis(((1_000_000f32 / output_refresh as f32) * 0.6f32) as u64);
```

Comment:
So the delay is longer, if the refresh-cycle is longer? That seems not very useful for very high refresh-rates, where we might select a too short duration and thus miss every other frame.

An optimal implementation would likely keep the last n rendering-times and do a sliding window average. That is overkill for anvil, but I feel like a bad example is worse than no example. Any ideas how we could make this less stupid without making it perfect?

Reply:
I choose this value cause I expect it to work fine for refresh rates up to 120Hz. For 120 Hz the compositor would have a bit more than 3 ms to render. My basic assumption for that was that if a user is able to drive such a high refresh rate the hardware would also be most likely capable of rendering the frame in a shorter period. On lower refresh rates a client would have more time to render. Like you said it is quite hard to select a reasonable delay. Even a sliding window with average has some limitations with spikes for example. I am not a fan of solely relying on some configuration parameter either because the topic may be hard to understand for an average user. On the other hand every hard coded value would be based on some specifc assumptions like the performance of the machine it was developed on. I would have skipped implementing the delay at all, but without it wp_presentation is rather useless imo as it would even increase the latency for some clients.

Currently I have no idea how that could be made better without introducing a dynamically calculated delay based on past render times and also some prediction.
But it definitely needs some comment describing why we want to delay the repaint.

Reply:
Seems reasonable. Once some docs are added, I would consider this resolved.

Reply:
done